x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/view/ClusterViewService.java

```diff
@@ -0,0 +1,119 @@
+
+    public ClusterViewService(ClusterService clusterService, ProjectResolver projectResolver, ViewServiceConfig config) {
+        super(config);
+        this.clusterService = clusterService;
+        this.projectResolver = projectResolver;
+        this.taskQueue = clusterService.createTaskQueue(
+            "update-esql-view-metadata",
+            Priority.NORMAL,
+            new SequentialAckingBatchedTaskExecutor<>()
+        );
+    }
+
+    @Override
+    protected ViewMetadata getMetadata() {
+        return getMetadata(projectResolver.getProjectId());
+    }
+
+    @Override
+    protected ViewMetadata getMetadata(ProjectId projectId) {
+        return getMetadata(clusterService.state().metadata().getProject(projectId));
+    }
+
+    protected ViewMetadata getMetadata(ProjectMetadata projectMetadata) {
+        return projectMetadata.custom(ViewMetadata.TYPE, ViewMetadata.EMPTY);
+    }
+
+    protected ProjectMetadata getProjectMetadata(ProjectId projectId) {
+        return clusterService.state().metadata().getProject(projectId);
+    }
+
+    @Override
+    protected void updateViewMetadata(
+        String verb,
+        ProjectId projectId,
+        AcknowledgedRequest<?> request,
+        ActionListener<? extends AcknowledgedResponse> callback,
+        Function<ViewMetadata, Map<String, View>> function
+    ) {
+        ViewMetadataUpdateTask updateTask = new ViewMetadataUpdateTask(request, callback, projectId, function);
+        String taskName = String.format(Locale.ROOT, "update-esql-view-metadata-[%s]", verb.toLowerCase(Locale.ROOT));
+        taskQueue.submitTask(taskName, updateTask, updateTask.timeout());
+    }
+
+    @SuppressForbidden(reason = "legacy usage of unbatched task") // TODO add support for batching here
+    private void submitUnbatchedTask(@SuppressWarnings("SameParameterValue") String source, ClusterStateUpdateTask task) {
+        clusterService.submitUnbatchedStateUpdateTask(source, task);
+    }
+
+    @Override
+    protected void assertMasterNode() {
+        assert clusterService.localNode().isMasterNode();
+    }
+
+    @Override
+    protected boolean viewsFeatureEnabled() {
+        return EsqlFeatures.ESQL_VIEWS_FEATURE_FLAG.isEnabled();
+    }
+
+    class ViewMetadataUpdateTask extends AckedClusterStateUpdateTask {
+        private final ProjectId projectId;
+        private final Function<ViewMetadata, Map<String, View>> updateFunction;
+
+        ViewMetadataUpdateTask(
+            AcknowledgedRequest<?> request,
+            ActionListener<? extends AcknowledgedResponse> listener,
+            ProjectId projectId,
+            Function<ViewMetadata, Map<String, View>> updateFunction
+        ) {
+            super(request, listener);
+            this.projectId = projectId;
+            this.updateFunction = updateFunction;
+        }
+
+        @Override
+        public ClusterState execute(ClusterState currentState) {
+            var project = getProjectMetadata(projectId);
```

Comment:
Using `getProjectMetadata(projectId)` here is wrong, as it gets the `ClusterState` from the `ClusterService` instead of using the one in this method. That means it will use an outdated cluster state to compute the updated cluster state. This is another reason why I'm not a huge fan of having a method like `getProjectMetadata` in this class, as it's easier to introduce bugs.